src/Utilities/CssDocument.php

```diff
@@ -82,23 +61,45 @@ public function __construct(string $css)
      */
     public function getStyleRulesData(array $allowedMediaTypes): array
     {
-        $splitCss = $this->splitCssAndMediaQuery($allowedMediaTypes);
-
         $ruleMatches = [];
-        foreach ($splitCss as $cssPart) {
-            // process each part for selectors and definitions
-            \preg_match_all('/(?:^|[\\s^{}]*)([^{]+){([^}]*)}/mi', $cssPart['css'], $matches, PREG_SET_ORDER);
-
-            foreach ($matches as $cssRule) {
+        /** @var CssRenderable $rule */
+        foreach ($this->sabberwormCssDocument->getContents() as $rule) {
+            if ($rule instanceof CssAtRuleBlockList) {
+                $media = $this->getFilteredAtIdentifierAndRule($rule, $allowedMediaTypes);
+                if ($media !== null) {
+                    /** @var CssRenderable $nestedRule */
+                    foreach ($rule->getContents() as $nestedRule) {
+                        if ($nestedRule instanceof CssDeclarationBlock) {
+                            $ruleMatches[] = [
+                                'media' => $media,
+                                'rule' => $nestedRule,
+                            ];
+                        }
+                    }
+                }
+            } elseif ($rule instanceof CssDeclarationBlock) {
                 $ruleMatches[] = [
-                    'media' => $cssPart['media'],
-                    'selectors' => $cssRule[1],
-                    'declarations' => $cssRule[2],
+                    'media' => '',
+                    'rule' => $rule,
                 ];
             }
         }
 
-        return $ruleMatches;
+        return \array_map(
+            /**
+             * @param array{media: string, rule: CssDeclarationBlock} $ruleMatch
+             *
+             * @return array{media: string, selectors: string, declarations: string}
+             */
+            static function (array $ruleMatch): array {
```

Comment:
Let's use a non-anonymous function instead: Then we can use its name to communicate its purpose.

Reply:
Done.  It's got to be a contender for the crappiest method I've ever written, and I hope will be gone when we replace these arrays with proper classes, but I'm all for atomizing code into the smallest constituent parts.